crates/red_knot_test/src/assertion.rs

```diff
@@ -0,0 +1,621 @@
+use ruff_python_trivia::CommentRanges;
+use ruff_source_file::{LineIndex, Locator, OneIndexed};
+use ruff_text_size::{Ranged, TextRange};
+use smallvec::SmallVec;
+use std::ops::Deref;
+
+/// Diagnostic assertion comments in a file.
+#[derive(Debug)]
+pub(crate) struct FileAssertions {
+    comment_ranges: CommentRanges,
+    source: SourceText,
+    lines: LineIndex,
+}
+
+impl FileAssertions {
+    pub(crate) fn from_file(db: &TestDb, file: File) -> Self {
+        let source = source_text(db, file);
+        let lines = line_index(db, file);
+        let parsed = parsed_module(db, file);
+        let comment_ranges = CommentRanges::from(parsed.tokens());
+        Self {
+            comment_ranges,
+            source,
+            lines,
+        }
+    }
+
+    fn line_number(&self, range: &impl Ranged) -> OneIndexed {
+        self.lines.line_index(range.start())
+    }
+
+    fn is_own_line(&self, range: &impl Ranged) -> bool {
+        CommentRanges::is_own_line(range.start(), &Locator::new(&self.source))
+    }
+}
+
+impl<'a> IntoIterator for &'a FileAssertions {
+    type Item = LineAssertions<'a>;
+    type IntoIter = LineAssertionsIterator<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        Self::IntoIter {
+            file_assertions: self,
+            inner: AssertionWithRangeIterator {
+                file_assertions: self,
+                inner: self.comment_ranges.into_iter(),
+            }
+            .peekable(),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct AssertionWithRange<'a>(Assertion<'a>, TextRange);
+
+#[derive(Debug)]
+struct AssertionWithRangeIterator<'a> {
+    file_assertions: &'a FileAssertions,
+    inner: std::iter::Copied<std::slice::Iter<'a, TextRange>>,
+}
+
+impl<'a> Iterator for AssertionWithRangeIterator<'a> {
+    type Item = AssertionWithRange<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let locator = Locator::new(&self.file_assertions.source);
+        loop {
+            let inner_next = self.inner.next()?;
+            let comment = locator.slice(inner_next);
+            if let Some(assertion) = Assertion::from_comment(comment) {
+                return Some(AssertionWithRange(assertion, inner_next));
+            };
+        }
+    }
+}
+
+impl std::iter::FusedIterator for AssertionWithRangeIterator<'_> {}
+
+impl<'a> Deref for AssertionWithRange<'a> {
+    type Target = Assertion<'a>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Ranged for AssertionWithRange<'_> {
+    fn range(&self) -> TextRange {
+        self.1
+    }
+}
+
+impl<'a> From<AssertionWithRange<'a>> for Assertion<'a> {
+    fn from(value: AssertionWithRange<'a>) -> Self {
+        value.0
+    }
+}
+
+/// A vector of [`Assertion`]s belonging to a single line.
+///
+/// Most lines will have zero or one assertion, so we use a [`SmallVec`] optimized for a single
+/// element to avoid most heap vector allocations.
+type AssertionVec<'a> = SmallVec<[Assertion<'a>; 1]>;
```

Comment:
That's a reasonable design, and maybe it would be simpler. At this point it feels like it would be a major rewrite of this module, and I'm not sure if that's the best use of my time; the current implementation works and I don't think it's so complex that it's a problem.

Reply:
With regards to dependencies -- I do actually agree that for a test framework specifically, local build times should possibly actually matter more than the time it actually takes to run the tests. And keeping dependencies down is one way to keep build times low. It's not (yet) nearly as big a problem in red-knot than it is in Ruff, but I find waiting for the tests to finish building far more frustrating when working on Ruff than waiting for the tests to complete once they've started.

Reply:
Does removing an external dependency that generally won't change between builds reduce build times significantly? I would assume on almost every build we do, smallvec will already be built and won't need rebuilding.

I do agree that build time matters here, though.

It's trivial to remove the smallvec dependency and just use a regular vec instead, but I think that could be a significant cost in test runtime (which I think will also matter, once we accumulate thousands of tests), if I don't do the full refactor to a flat-vector-with-ranges.

Reply:
> Does removing an external dependency that generally won't change between builds reduce build times significantly? I would assume on almost every build we do, smallvec will already be built and won't need rebuilding.

That's true. After the weekly renovate PRs on Mondays a local build generally takes a while longer, and that issue gets worse the more dependencies you have. But that's generally only a weekly issue, so I suppose that's not really too bad either.

Reply:
I'm not too concerned about the smallvec dependency. My main concern is that iterators result in a lot of boilerplate and lifetimes that can be avoided by collecting to a vector. The current design also requires an optimization (small vec) that otherwise wouldn't be necessary. I suspect that it's slightly more code than necessary and refactoring would only require moving some logic around (but not rewriting). 

I do agree with that this might not be the best use of your time to refactor it now. We could consider this a starter task for David, leave it as is and add a TODO, or you try how far you get by spending a limited time on it.

Reply:
I don't think I'm going to do this for assertions in this PR. If someone is motivated to do it in another PR, that's fine, but I'm not really convinced there are strong reasons to do it at all. Based on doing this transformation for diagnostics, I don't see evidence that it will make the code any shorter or simpler. I think it was worth it for diagnostics because we will often have two per line, reducing the effectiveness of the smallvec; I don't think that's true here.